runner/copy.go

```diff
@@ -258,39 +282,64 @@ func (m *MaterializedChunkRunnerConfig) Deserialize(config Config) error {
 	return nil
 }
 
+// We'll often build many MaterializedChunkRunners with the same provider configs, this is an optimization
+// to avoid that. This assumes the providers are thread-safe however. We separate the caches because some
+// tests will create empty configs for both an online and offline store and they'll map to the same thing.
+var onlineProviderCache = &sync.Map{}
+var offlineProviderCache = &sync.Map{}
+
 func MaterializedChunkRunnerFactory(config Config) (types.Runner, error) {
 	runnerConfig := &MaterializedChunkRunnerConfig{}
 	if err := runnerConfig.Deserialize(config); err != nil {
 		return nil, err
 	}
+	var onlineStore provider.OnlineStore
+	var offlineStore provider.OfflineStore
+	onlineCfg := runnerConfig.OnlineConfig
+	offlineCfg := runnerConfig.OfflineConfig
+	onlineCfgStr := string(onlineCfg)
+	offlineCfgStr := string(offlineCfg)
 
-	onlineProvider, err := provider.Get(runnerConfig.OnlineType, runnerConfig.OnlineConfig)
-	if err != nil {
-		return nil, err
-	}
-	offlineProvider, err := provider.Get(runnerConfig.OfflineType, runnerConfig.OfflineConfig)
-	if err != nil {
-		return nil, err
-	}
-	onlineStore, err := onlineProvider.AsOnlineStore()
-	if err != nil {
-		return nil, err
+	cachedOnline, found := onlineProviderCache.Load(onlineCfgStr)
```

Comment:
I wonder if it makse sense to move the caching logic into provider.Get

Reply:
@sdreyer did some optimization work for redis I wonder if any of that work applies

Reply:
I actually turned this off, because it breaks all tests and has really gnarly edge cases where you can get two different providers with the same config